update variables docs (#725)
okteto/docs: src/content/core/okteto-variables.mdx

import Image from "@theme/Image";

Configuration is a core concept of every application, we understand how important it is for developers and teams to parametrize their project deployments and Okteto manifests. At Okteto we built different ways for you and your teams to manage
and share config values using Okteto Variables.

> This page is a guide for those who are authoring Okteto manifests and configuring Okteto for their team and organization. If you are interested in runtime variables of dev environments, see [Development Containers](/docs/development/containers).

Reviewer: ```suggestion
This page is a guide for those who are authoring Okteto manifests and configuring Okteto for their team and organization. If you are interested in runtime variables of dev environments, see [Development Containers](development/containers/index.mdx).
```

## What are Okteto Variables?

> Okteto Variables are a way to parametrize your [Okteto manifest](/docs/reference/okteto-manifest) and the deploy, destroy and test strategy.

Reviewer: ```suggestion
Okteto Variables are a way to parametrize your [Okteto manifest](reference/okteto-manifest.mdx) and the deploy, destroy and test strategy.
```

- **Catalog Variables**: Used to define application-specific variables
- **Local Environment Variables**: Used to define variables ad-hoc when using the Okteto CLI
- **The `.env` file**: Used to define project or team-specific variables
- **User Variables**: Used to define variables at the user level
- **Admin Variables**: Used to define variables for all users in the organization

:::note

The above listing order is also the order in which the Okteto CLI gives precedence to each variable type. At the top the variables with the highest precedence.

Variables with the same name defined in different types will be overridden by the variable with higher precedence.

Example: the var `PORT` defined both in `.env` as `PORT=3000` and as `okteto deploy --var PORT=4000`, the value of `PORT` will be `4000` because the `--var` flag has higher precedence.

:::

### Scope

Okteto Variables are scoped to the execution of the Okteto CLI. This means that they are:

- available for expanding values in the Okteto manifest
- available to commands, scripts and tools running as part of `deploy`, `destroy` and `test` commands in the Okteto manifest

> Okteto Variables **are not** propagated to the development containers. If you need to pass variables to your development containers, see [Development Containers](/docs/development/containers).

Reviewer: ```suggestion
Okteto Variables **are not** propagated to the development containers. If you need to pass variables to your development containers, see [Development Containers](development/containers/index.mdx).
```

```yaml
deploy:
  commands:
  - name: Create AWS infrastructure
    command: |
      set -e
      bucketName="${OKTETO_NAMESPACE}-okteto-tacos-shop"
      terraform apply -input=false -var "bucket_name=$bucketName" -auto-approve
      echo "S3_BUCKET_NAME=$resourceName" >> "$OKTETO_ENV"

  - name: Create .env
    command: |
      set -e
      echo "S3_BUCKET_NAME=$S3_BUCKET_NAME" > .env

  - name: Deploy the app
    command: |
      set -e
      make deploy
```

:::note

Here are some highlights, and clarifications on the example above:

1. The first step makes the variable `S3_BUCKET_NAME` available to the rest of the deployment steps. Both the step "Create .env" and "Deploy the app" will have access to `S3_BUCKET_NAME`.

2. Also, it's important to remember that the variables in `$OKTETO_ENV` are not shared across `deploy`, `test`, and `destroy` sections. They are only shared within the same section.

3. If you are wondering where the variable `OKTETO_NAMESPACE` comes from, check the section Okteto Variables [Built-in by Okteto](#built-in-by-okteto).

:::

### Configuring dynamic endpoints

> In this example, we show how `$OKTETO_ENV` can be used to configure dynamic endpoints for [External Resources](/docs/core/okteto-manifest/#external) within your dev environment.

Reviewer: ```suggestion
In this example, we show how `$OKTETO_ENV` can be used to configure dynamic endpoints for [External Resources](core/okteto-manifest.mdx#external) within your dev environment.
```

import Image from "@theme/Image";

Configuration is a core concept of every application, we understand how important it is for developers and teams to parametrize their project deployments and Okteto manifests. At Okteto we built different ways for you and your teams to manage
and share config values using Okteto Variables.

This page is a guide for those who are authoring Okteto manifests and configuring Okteto for their team and organization. If you are interested in runtime variables of dev environments, see [Development Containers](/docs/development/containers).

## What are Okteto Variables?

Okteto Variables are a way to parametrize your [Okteto manifest](/docs/reference/okteto-manifest) and the deploy, destroy and test strategy.

### Types of Variables

Each type is designed to solve a specific use-case.

> - **Managed by Okteto**: variables injected in your dev environment automatically by Okteto

Reviewer: I would use the term `Built-in by Okteto`here

Author: agree, managed sounds something different, updated in f5a0f5ca3cf50bd567be6678b0fca19ddd54d0ec

> - **The `--var` flag**: Used to define variables ad-hoc when using the Okteto CLI

Reviewer: I would use something "variables defined at deployment time". It can be from the CLI or using the Okteto UI / deploy dialog

Author: good shout, changed to:

- **Deployment Variables**: Used to define variables at deployment time, either from the Okteto CLI or the Okteto UI deploy dialog

> - **Catalog**: Used to define project-specific variables

Reviewer: `variables defined in the Catalog`? -> we use the term "applications" when talking about catalog:
https://www.okteto.com/docs/admin/catalog/

Author: Iterated on it, changed to:

- **Catalog Variables**: Used to define application-specific variables

- **Local Environment Variables**: Used to define variables ad-hoc when using the Okteto CLI
- **The `.env` file**: Used to define project or team-specific variables

> - **User Variables**: Used to define variables at the user level

Reviewer: User-defined variables? and maybe something like "variables defined by each user than are injected in the dev environments of this user"

Author: I think all types of vars are 'defined' in some way, so I didn't specify it to any type, but we can change it for User and Admin if it helps passing the concept, personally I think it complicates the reading, but it's maybe personal bias. What do others prefer/think? cc @rlamana @codyjlandstrom

Author: also, I think I made the mistake of mentioning "injected in dev environment" which I will remove, because at the top I make it very clear that this doc is for manifest expansion and deploy/destroy/test variables, not vars that end up in the dev container.

Reviewer: I like it a lot how it's done now :-)

> - **Admin Variables**: Used to define variables for all users in the organization

Reviewer: Admin-defined variables? and maybe something like "variables defined by Okteto Admins than are injected in the dev environments of every user"

:::note

The above listing order is also the order in which the Okteto CLI gives precedence to each variable type. At the top the variables with the highest precedence.

Variables with the same name defined in different types will be overridden by the variable with higher precedence.

Example: the var `PORT` defined both in `.env` as `PORT=3000` and as `okteto deploy --var PORT=4000`, the value of `PORT` will be `4000` because the `--var` flag has higher precedence.

:::

### Scope

Okteto Variables are scoped to the execution of the Okteto CLI. This means that they are:

- available for expanding values in the Okteto manifest

> - available in the `deploy`, `destroy` and `test` commands in the Okteto manifest

Reviewer: I think this one is redundant

Author: do you mean because the first item already covers it?

I used a slighly redundant writing to make sure to cover all cases and remove ambiguity.

Let me know how you would change it. Remove or incorporate into a single item? @pchico83

Reviewer: I think the second one is redundant with the third one

Author: you're right, too redundant. Made this change:

```
-- available in the `deploy`, `destroy` and `test` commands in the Okteto manifest
-- available to scripts and tools running as part of `deploy`, `destroy` and `test` commands in the Okteto manifest
+- available to commands, scripts and tools running as part of `deploy`, `destroy` and `test` commands in the Okteto manifest
```

Perhaps it's obvious, I hope it's clear that Okteto Variables are available both in commands but also internet tooling invoked by such commands.

- available to scripts and tools running as part of `deploy`, `destroy` and `test` commands in the Okteto manifest

> They **are not** propagated to the development containers. If you need to pass variables to your development containers, see [Development Containers](/docs/development/containers).

Reviewer: ```suggestion
Okteto Variables **are not** propagated to the development containers. If you need to pass variables to your development containers, see [Development Containers](/docs/development/containers).
```

:::note

> Variables defined in a specific command item are only scoped to that item by default. To share values between command items, use the special variable [$OKTETO_ENV](#sharing-variables-between-commands-with-okteto_env).

Reviewer: There is a section later about `OKTETO_ENV`. I think this is redundant

Example:

```yaml
deploy:
  commands:
    - export MY_VAR=hello
    - echo $MY_VAR # -> ''
    - echo "MY_VAR=hello" >> $OKTETO_ENV
    - echo $MY_VAR # -> 'hello'
```

:::

## Use Cases

### Expand values in the Okteto manifest

Variables can be used to **expand values** in the the **Okteto manifest** (see `${PORT}`)

```yaml
# okteto.yml
dev:
  api:
    command: npm start
    environment:
      PORT: ${PORT}
```

### Deploy, destroy and test commands

Variables can be used in **deploy**, **destroy** and **test** commands in the Okteto manifest (see `${ENV}`)

> ```yaml

Reviewer: I think this sample is redundant

Author: Agree, I want to make sure everything is still mentioned explicitly, that commands and tools/scripts have access to Okteto Variables, but I've condensated the 2 pagragraphs into 1 example.

<img width="1046" alt="Screenshot 2024-05-10 at 16 28 52" src="https://github.com/okteto/docs/assets/2318450/bdc8d102-13cd-4137-a44d-842d64e8fe38">

Happy to iterate further if anyone has feedback 🙏

# okteto.yml
deploy:
  commands:
    - terraform init -backend-config=./config/${ENV}/backend.hcl
    - terraform plan -var-file=./config/${ENV}/terraform.tfvars -out tfplan
    - terraform apply tfplan
```

> Additionally, variables are injected as environment variable of the Okteto CLI so your tools or scripts can use them

Reviewer: ```suggestion
Additionally, Okteto Variables are injected as environment variable of the Okteto CLI so your tools or scripts can use them
```

```yaml
test:
  commands:
    # Example: some infrastructure has been provisioned
    - terraform init -backend-config=./config/${ENV}/backend.hcl
    - terraform plan -var-file=./config/${ENV}/terraform.tfvars -out tfplan
    - terraform apply tfplan
    - terraform output -json > tfout.json

    # I make the following var available to other steps
    - SERVICE_ENDPOINT=$(cat tfout.json | jq -r '.service_endpoint') >> $OKTETO_ENV

    # the following step will have access to SERVICE_ENDPOINT
    - echo "Running tests against ${SERVICE_ENDPOINT}..."
```

Another example for the same functionality, can be for example, creating the infrastructure, generating a `.env` file, and then deploy the application.

```yaml
deploy:
  commands:
  - name: Create AWS infrastructure
    command: |
      set -e
      bucketName="${OKTETO_NAMESPACE}-okteto-tacos-shop"
      terraform apply -input=false -var "bucket_name=$bucketName" -auto-approve
      echo "S3_BUCKET_NAME=$resourceName" >> "$OKTETO_ENV"

  - name: Create .env
    command: |
      set -e
      echo "S3_BUCKET_NAME=$S3_BUCKET_NAME" > .env

  - name: Deploy the app
    command: |
      set -e
      make deploy
```

:::note

> Here are some highlights, and clarifications on the example above:

Reviewer: I would also add a sample using a multi-line command. Multiline commands can be very convenient to avoid the usage of `OKTETO_ENV`.

Author: Agree, I've added it 👍

1. The first step makes the variable `S3_BUCKET_NAME` available to the rest of the deployment steps. Both the step "Create .env" and "Deploy the app" will have access to `S3_BUCKET_NAME`.

2. Also, it's important to remember that the variables in `$OKTETO_ENV` are not shared across `deploy`, `test`, and `destroy` sections. They are only shared within the same section.

3. If you are wondering where the variable `OKTETO_NAMESPACE` comes from, check the section Okteto Variables [Managed by Okteto](#managed-by-okteto).

:::

### Configuring dynamic endpoints

> In this example, we show how `$OKTETO_ENV` can be used to create dynamic endpoints for external resources within your dev environment.

Reviewer: ```suggestion
In this example, we show how `$OKTETO_ENV` can be used to create dynamic endpoints for External Resources within your dev environment.
```
Also, add a link to https://www.okteto.com/docs/core/okteto-manifest/#external

> If you are not familiar yet with the concept of Okteto external resource, image a component that is not part of the deployment of your dev environment, but that you need to interact with. For example, a database, a message queue, or an S3 bucket. If you need more details on this feature, check out [External Resources](/docs/development/external-resources).

Reviewer: I think we can skip this paragraphn with the previous one if we link to https://www.okteto.com/docs/core/okteto-manifest/#external before

> External resources are defined in the Okteto manifest and have a name and a URL. We understand that often, when using cloud providers, the resource URLs are generated dynamically when the infrastructure is provisioned. In this example, we make the S3 bucket name available to the rest of the deployment steps.

Reviewer: ```suggestion
External Resources are defined in the Okteto manifest and have a name and a URL. We understand that often, when using cloud providers, the resource URLs are generated dynamically when the infrastructure is provisioned. In this example, we make the S3 bucket name available to the rest of the deployment steps.
```

```yaml
## TODO: write example
```

## Setting Okteto Variables

This section will show a brief example of how to set each type of Okteto Variable.

> ### The --var flag

Reviewer: LEt's use the same terms as in the intro

The `--var` flag is useful to specify ad-hoc variables when running a command like [deploy](/docs/reference/okteto-cli#deploy) or [preview deploy](/docs/reference/okteto-cli#deploy-3)

Example:

```console
$ okteto deploy --var PORT=4000
```

```yaml
deploy:
  image: "node:${TAG}"
  commands:
    - name: deploy
      command: NODE_ENV=${ENV} make deploy
```

### Catalog

> When you deploy an application from the Okteto Dashboard or if your organization is using the Okteto Catalog, you can define project-specific variables. See [Catalog Variables](/docs/admin/catalog/#add-environment-variables) for more information.

Reviewer: ```suggestion
When you deploy an application from the Okteto UI or if your organization uses the Okteto Catalog, you can define variables specific for your development environment. See [Catalog Variables](/docs/admin/catalog/#add-environment-variables) for more information.
```
We don't use the term "project" in our docs. I know it's hard, and that our terms might not be ideal, but it's better to stick to them than adding new synonymous

### Local Environment Variables

Similar to the `--var` flag, local environment variables are useful when you need to specify ad-hoc variables when running a command like [deploy](/docs/reference/okteto-cli#deploy) or [preview deploy](/docs/reference/okteto-cli#deploy-3)

Example:

```console
$ PORT=4000 okteto deploy
```

### The .env file

> You can set default values for any environment variables referenced in the okteto manifest in an environment file named `.env`.

Reviewer: ```suggestion
You can set default values for any environment variables referenced in the Okteto Manifest in an environment file named `.env`.
```

The `.env` should be placed in the same folder of the Okteto Manifest. The Okteto Manifest can also be stored in a special folder called `.okteto`. In that case, the `.env` should not be placed inside it, but in the parent directory.

For example:

```yaml
# .env
PORT=4000
```

```yaml
# okteto.yml
deploy:
  commands:
    - name: deploy
      command: echo $PORT # -> 4000
```

:::tip

The `.env` integration also supports parameter expansion to set default values and allowing other developers to override them.

Example:

```yaml
# .env
PORT=${PORT:-4000}
```

When running `PORT=5000 okteto deploy` the value of `PORT` will be `5000`.

:::

> ### User and Admin variables

Reviewer: They are different and created from different places, I would split this section in two

Author: agree, already changed 👍

This page is a guide for those who are authoring Okteto manifests and configuring Okteto for their team and organization. If you are interested in runtime variables of dev environments, see [Development Containers](/docs/development/containers).

## What are Okteto Variables?

Okteto Variables are a way to parametrize your [Okteto manifest](/docs/reference/okteto-manifest) and deploy, destroy and test strategy.

### Types of Variables

Each type is designed to solve a specific use-case.

- **Managed by Okteto**: variables injected in your dev environment automatically by Okteto, see [Default Environment Variables](#default-environment-variables) and [Built-in Environment Variables for Images in Okteto Registry](#built-in-environment-variables-for-images-in-okteto-registry).
- **The `--var` flag**: Used to define variables ad-hoc when using the Okteto CLI
- **Catalog**: Used to define project-specific variables
- **Local Environment Variables**: Used to define variables ad-hoc when using the Okteto CLI
- **The `.env` file**: Used to define project or team-specific variables
- **User Variables**: Used to define variables at the user level

> - **Admin Variables**: Used to define variables at the organization level

Reviewer 2: Not sure "at the organization level" is clear enough and users will understand what it implies. What about something like this?  
```Defined by admins, these variables will be available(or injected?) to all users at the organization level```
cc. @codyjlandstrom

Author: to keep it consisent with the rest of the items (short an concise, straight to the point), could be: `Used to define variables for all users of the organization`. WDYT? cc @codyjlandstrom @rlamana

:::note

The above listing order is also the order in which the Okteto CLI gives precedence to each variable type. At the top the variables with the highest precedence.

Variables with the same name defined in different types will be overridden by the variable with higher precedence.

Example: the var `PORT` defined both in `.env` as `PORT=3000` and as `okteto deploy --var PORT=4000`, the value of `PORT` will be `4000` because the `--var` flag has higher precedence.

:::

### Scope

Okteto Variables are scoped to the execution of the Okteto CLI. This means that they are:

- available for expanding values in the Okteto manifest
- available in the `deploy`, `destroy` and `test` commands in the Okteto manifest
- available to scripts and tools running as part of `deploy`, `destroy` and `test` commands in the Okteto manifest

They **are not** propagated to the development containers. If you need to pass variables to your development containers, see [Development Containers](/docs/development/containers).

:::note

> Variables defined in a specific command item are only scoped to that item by default. To share valyes between command items, use the special variable [$OKTETO_ENV](#sharing-variables-between-commands-with-okteto_env).

Reviewer: ```suggestion
Variables defined in a specific command item are only scoped to that item by default. To share values between command items, use the special variable [$OKTETO_ENV](#sharing-variables-between-commands-with-okteto_env).
```

```
- Variables cab be used in `deploy`, `destroy` and `test` commands in the Okteto manifest

### Catalog

When you deploy an application from the Okteto Dashboard or if your organization is using the Okteto Catalog, you can define project-specific variables. See [Catalog Variables](/docs/admin/catalog/#add-environment-variables) for more information.

### Local Environment Variables

Similar to the `--var` flag, local environment variables are useful when you need to specify ad-hoc variables when running a command like [deploy](/docs/reference/okteto-cli#deploy) or [preview deploy](/docs/reference/okteto-cli#deploy-3)

Example:

```console
$ PORT=4000 okteto deploy
```

### The .env file

You can set default values for any environment variables referenced in the okteto manifest in an environment file named `.env`.
The `.env` file is placed at the same folder than the okteto manifest.

For example:

```console
$ cat .env
TAG=v1.5
```

```console
$ cat okteto.yaml
...
    image: "node:${TAG}"
...
```

When you run any Okteto CLI command, `image` will be `node:v1.5`.

> Values in the shell take precedence over those specified in the `.env` file.

### User and Admin variables

> You can create and delete your Okteto Variables from the `Variables` section in the `Settings` view of the Okteto dashboard:

Reviewer 2: We might need to rephrase this and update screenshots according to this PR here: https://github.com/okteto/app/pull/8040 for Admin Variables

:::

## Setting Okteto Variables

This section will show a brief example of how to set each type of Okteto Variable.

### Deployment Variables

> Variables defined at deployment time are those set with the `--var` flag. They are used to specify ad-hoc variables when running a command like [deploy](/docs/reference/okteto-cli#deploy) or [preview deploy](/docs/reference/okteto-cli#deploy-3)

Reviewer: ```suggestion
Variables defined at deployment time are those set with the `--var` flag. They are used to specify ad-hoc variables when running a command like [deploy](reference/okteto-cli.mdx#deploy) or [preview deploy](reference/okteto-cli.mdx#deploy-3)
```

Example:

```console
$ okteto deploy --var PORT=4000
```

```yaml
deploy:
  image: "node:${TAG}"
  commands:
    - name: deploy
      command: NODE_ENV=${ENV} make deploy
```

### Catalog Variables

> When you deploy an application from the Okteto UI or if your organization uses the Okteto Catalog, you can define variables specific for your development environment. See [Catalog Variables](/docs/admin/catalog/#add-environment-variables) for more information.

Reviewer: ```suggestion
When you deploy an application from the Okteto UI or if your organization uses the Okteto Catalog, you can define variables specific for your development environment. See [Catalog Variables](admin/catalog.mdx#add-environment-variables) for more information.
```

  />
</p>

## Managed by Okteto

> ### Default Environment Variables

Reviewer: This should be the first ones, to follow the same order than in the intro paragraph

Author: moved above 👍
